doxygen/man/man3/AddressTableModel.3
.TH "AddressTableModel" 3 "Wed Feb 10 2016" "Version 1.0.0.0" "darksilk" \" -*- nroff -*-
.ad l
.nh
.SH NAME
AddressTableModel \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <addresstablemodel\&.h>\fP
.PP
Inherits QAbstractTableModel\&.
.SS "Public Types"

.in +1c
.ti -1c
.RI "enum \fBAddressType\fP { \fBAT_Unknown\fP = 0, \fBAT_Normal\fP = 1, \fBAT_Stealth\fP = 2 }"
.br
.ti -1c
.RI "enum \fBColumnIndex\fP { \fBLabel\fP = 0, \fBAddress\fP = 1, \fBType\fP = 2 }"
.br
.ti -1c
.RI "enum \fBRoleIndex\fP { \fBTypeRole\fP = Qt::UserRole }"
.br
.ti -1c
.RI "enum \fBEditStatus\fP { \fBOK\fP, \fBNO_CHANGES\fP, \fBINVALID_ADDRESS\fP, \fBDUPLICATE_ADDRESS\fP, \fBWALLET_UNLOCK_FAILURE\fP, \fBKEY_GENERATION_FAILURE\fP }"
.br
.in -1c
.SS "Public Slots"

.in +1c
.ti -1c
.RI "void \fBupdateEntry\fP (const QString &address, const QString &label, bool isMine, int status)"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBAddressTableModel\fP (\fBCWallet\fP *wallet, \fBWalletModel\fP *parent=0)"
.br
.ti -1c
.RI "\fB~AddressTableModel\fP ()"
.br
.ti -1c
.RI "QString \fBaddRow\fP (const QString &type, const QString &label, const QString &address, int addressType)"
.br
.ti -1c
.RI "QString \fBlabelForAddress\fP (const QString &address) const "
.br
.ti -1c
.RI "int \fBlookupAddress\fP (const QString &address) const "
.br
.ti -1c
.RI "\fBEditStatus\fP \fBgetEditStatus\fP () const "
.br
.in -1c
.PP
.RI "\fBMethods overridden from QAbstractTableModel\fP"
.br

.in +1c
.in +1c
.ti -1c
.RI "int \fBrowCount\fP (const QModelIndex &parent) const "
.br
.ti -1c
.RI "int \fBcolumnCount\fP (const QModelIndex &parent) const "
.br
.ti -1c
.RI "QVariant \fBdata\fP (const QModelIndex &\fBindex\fP, int role) const "
.br
.ti -1c
.RI "bool \fBsetData\fP (const QModelIndex &\fBindex\fP, const QVariant &\fBvalue\fP, int role)"
.br
.ti -1c
.RI "QVariant \fBheaderData\fP (int section, Qt::Orientation orientation, int role) const "
.br
.ti -1c
.RI "QModelIndex \fBindex\fP (int row, int column, const QModelIndex &parent) const "
.br
.ti -1c
.RI "bool \fBremoveRows\fP (int row, int count, const QModelIndex &parent=QModelIndex())"
.br
.ti -1c
.RI "Qt::ItemFlags \fBflags\fP (const QModelIndex &\fBindex\fP) const "
.br
.in -1c
.in -1c
.SS "Static Public Attributes"

.in +1c
.ti -1c
.RI "static const QString \fBSend\fP = 'S'"
.br
.ti -1c
.RI "static const QString \fBReceive\fP = 'R'"
.br
.in -1c
.SS "Friends"

.in +1c
.ti -1c
.RI "class \fBAddressTablePriv\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
Qt model of the address book in the core\&. This allows views to access and modify the address book\&. 
.PP
Definition at line 14 of file addresstablemodel\&.h\&.
.SH "Member Enumeration Documentation"
.PP 
.SS "enum \fBAddressTableModel::AddressType\fP"

.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fIAT_Unknown \fP\fP
User specified label 
.TP
\fB\fIAT_Normal \fP\fP
DarkSilk address 
.TP
\fB\fIAT_Stealth \fP\fP
Stealth address 
.PP
Definition at line 22 of file addresstablemodel\&.h\&.
.SS "enum \fBAddressTableModel::ColumnIndex\fP"

.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fILabel \fP\fP
User specified label 
.TP
\fB\fIAddress \fP\fP
DarkSilk address 
.TP
\fB\fIType \fP\fP
Address type 
.PP
Definition at line 28 of file addresstablemodel\&.h\&.
.SS "enum \fBAddressTableModel::EditStatus\fP"
Return status of edit/insert operation 
.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fIOK \fP\fP
Everything ok 
.TP
\fB\fINO_CHANGES \fP\fP
No changes were made during edit operation 
.TP
\fB\fIINVALID_ADDRESS \fP\fP
Unparseable address 
.TP
\fB\fIDUPLICATE_ADDRESS \fP\fP
Address already in address book 
.TP
\fB\fIWALLET_UNLOCK_FAILURE \fP\fP
Wallet could not be unlocked to create new receiving address 
.TP
\fB\fIKEY_GENERATION_FAILURE \fP\fP
Generating a new public key for a receiving address failed 
.PP
Definition at line 39 of file addresstablemodel\&.h\&.
.SS "enum \fBAddressTableModel::RoleIndex\fP"

.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fITypeRole \fP\fP
Type of address (\fBSend\fP or \fBReceive\fP) 
.PP
Definition at line 34 of file addresstablemodel\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "AddressTableModel::AddressTableModel (\fBCWallet\fP * wallet, \fBWalletModel\fP * parent = \fC0\fP)\fC [explicit]\fP"

.PP
Definition at line 153 of file addresstablemodel\&.cpp\&.
.SS "AddressTableModel::~AddressTableModel ()"

.PP
Definition at line 161 of file addresstablemodel\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "QString AddressTableModel::addRow (const QString & type, const QString & label, const QString & address, int addressType)"

.PP
Definition at line 349 of file addresstablemodel\&.cpp\&.
.SS "int AddressTableModel::columnCount (const QModelIndex & parent) const"

.PP
Definition at line 172 of file addresstablemodel\&.cpp\&.
.SS "QVariant AddressTableModel::data (const QModelIndex & index, int role) const"

.PP
Definition at line 178 of file addresstablemodel\&.cpp\&.
.SS "Qt::ItemFlags AddressTableModel::flags (const QModelIndex & index) const"

.PP
Definition at line 312 of file addresstablemodel\&.cpp\&.
.SS "\fBEditStatus\fP AddressTableModel::getEditStatus () const\fC [inline]\fP"

.PP
Definition at line 77 of file addresstablemodel\&.h\&.
.SS "QVariant AddressTableModel::headerData (int section, Qt::Orientation orientation, int role) const"

.PP
Definition at line 300 of file addresstablemodel\&.cpp\&.
.SS "QModelIndex AddressTableModel::index (int row, int column, const QModelIndex & parent) const"

.PP
Definition at line 329 of file addresstablemodel\&.cpp\&.
.SS "QString AddressTableModel::labelForAddress (const QString & address) const"

.PP
Definition at line 468 of file addresstablemodel\&.cpp\&.
.SS "int AddressTableModel::lookupAddress (const QString & address) const"

.PP
Definition at line 499 of file addresstablemodel\&.cpp\&.
.SS "bool AddressTableModel::removeRows (int row, int count, const QModelIndex & parent = \fCQModelIndex()\fP)"

.PP
Definition at line 449 of file addresstablemodel\&.cpp\&.
.SS "int AddressTableModel::rowCount (const QModelIndex & parent) const"

.PP
Definition at line 166 of file addresstablemodel\&.cpp\&.
.SS "bool AddressTableModel::setData (const QModelIndex & index, const QVariant & value, int role)"

.PP
Definition at line 225 of file addresstablemodel\&.cpp\&.
.SS "void AddressTableModel::updateEntry (const QString & address, const QString & label, bool isMine, int status)\fC [slot]\fP"

.PP
Definition at line 343 of file addresstablemodel\&.cpp\&.
.SH "Friends And Related Function Documentation"
.PP 
.SS "friend class \fBAddressTablePriv\fP\fC [friend]\fP"

.PP
Definition at line 94 of file addresstablemodel\&.h\&.
.SH "Member Data Documentation"
.PP 
.SS "const QString AddressTableModel::Receive = 'R'\fC [static]\fP"
Specifies receive address 
.PP
Definition at line 49 of file addresstablemodel\&.h\&.
.SS "const QString AddressTableModel::Send = 'S'\fC [static]\fP"
Specifies send address 
.PP
Definition at line 48 of file addresstablemodel\&.h\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for darksilk from the source code\&.


CHANGELOG
Base58 Change, Port Changes and Minor Fixes
Addition of OBJ directory for Daemon builds
Version updated
ReadMe updated
Instamine code for Silk/Dark weaver swap/burn added to main.cpp
OpenSSL update
LevelDB 1.18 Update
Fix to guard against OpenSSL's new strict DER checks, fixed getbalance and other minor fixes
Spelling edits
Enabled locking of Stormnodes via the Stormnode configuration file
Unset Sandstorm checkbox if there were non-anonymized inputs selected in coincontrol, give warning
Removal of target space change and hardfork
Fixed time to switch "out of sync" messages
Fix twice denominatedBalance calculations
Solve performance problems by caching and fix probable security issue
RPC: Allow "sandstorm" only with ENABLE_WALLET
Disconnect Sandstorm timer on Overview page destruction
Fix spaces/monospace fonts in qt console
Mac compatibility so users can switch tabs using CMD+(NUM)
Added better instructions to get libsecp256k1 installed in case of library issues within ReadMe
Added secp256k1 libs to /deps in case for any reason the libs are not available from github anymore
Increased maximum block size to 50MB (50x Bitcoin Core)
Added safe shutdown for Windows
Fixed sync issues and CPU usage
Add event support for Tor
rpc: Prevent easy memory exhaustion attack 
Fixed issue where using setaccount with a foreign address causes the address to be added to your receiving addresses.
Fix priority calculation in CreateTransaction
darksilk.conf added to /src
GUI Styling changes
Statistics Page Added and Icons Updated
Clickable Padlock to Unlock Wallet and Unlock for Staking Only
Fixed GUI bugs
Assertion Error Fix for MacOSX
Sandstorm GUI Fixes
Decentralised DarkSilkMarket added
Overview Transactions UI Fixes
Messaging, GUI Update, Null Pointer Fix, Various Fixes/Changes
Windows Build Fix
Addition of possibly needed dependency
Stormnode Collateral Increase
Marketplace RPC Calls | Cleanup
Amended Spork Max Value for Stormnode Collateral
Removal of unused Spork
Minimum Transaction fee changed to 0.0001DRKSLK
Add support for miniupnpc API version 14
DARKSILK_TIMEDATA_MAX_SAMPLES 200
Add time offset to getpeerinfo output
Don't check duplicate stake while importing
Ignore blocks received while importing
Further Market RPC JSON calls
marketbuyrequest RPC command
marketmybuys RPC call for market
MacOSX Fixes and GUI Fix
Market RPC Minor Fixes
darksilk-qt.pro cleanup
DarkSilk-qt.app renamed to DarkSilk.app for OSX
Replaced stake min age with min confirmation
Limited drop of difficulty per block
Made special marker non spendable
Added nonspendable encoding for block signing key
BIP66 compatibility
Replaced NOP7 with CHECKLOCKTIMEVERIFY
Enforce new script verification flags
Static pos reward
Demand canonical block sig from updated peers
Submitblock: Add submission of stake template
Reward and Min Conf Blocks Changed
Checkpointing Removal
Gitian MacDeploy Graphics Added
IsProtocolV1/IsProtocolV2 hard fork switches removed
FutureDriftV1/FutureDriftV2 reverted back to FutureDrift
CheckStakeKernelHashV1/CheckStakeKernalHashV2 reverted back to CheckStakeKernelHash
Unsigned Int GetTargetSpacing changed to Static Const Int POS_TARGET_SPACING
Fix boost::get usage with Boost 1.58 in rpcrawtransaction.cpp 
Static PoS Reward of 1 DRKSLK
Prevent redefinition compiler warning
Fix issue where getrawtransaction RPC call does not work
PoW additions to GetNextTargetRequired
Scale up addrman
Gitian windows compile fixes
Correctly randomize change output position
LevelDB windows cross compile missing files
Scale up addrman
Gitian windows compile fixes
Correctly randomize change output position
Removal of Dead Code
Fix for building with Boost 1.58
LevelDB Cross compile fixes
DarkSilk-Qt: cleanup / optimise addressbookpage
Discourage fee sniping with nLockTime
Removal of dead code from main.h
Changed ProofOfStakeReward to use static const
Diff changes
Shadowchat Messaging Updates
DNSSeed Thread Fix
Stormnode payments start on block 82002
Restore hex to getrawtransaction vout scriptPubkey
static const rename | nActualSpacing code reform
Removal of Keepass until later date that this doesn't cause an issue with Windows builds
Updated dependencies, added Gitian OSX files
Remove unused GetProofOfStakeReward
Faster Retargetting for PoW
Fixed PoS target spacing
Fix payee selectoin on StormnodePayments::ProcessBlock
fix log output / more notCapableReason-s
HasCollateralInputs should check only for at least 1 collateral not 2
Track lastTimeChanged for mnodeman and ask for update only if UpdateNeeded
Added sstx support for free transactions 
Fix for src/qt/stormnodemanager.cpp
RPC Fixes
Fix DoAutomaticDenominating
Few Fixes for Processing Extra Messages
Reversioning and restriction change
Check and remove expired Stormnodes on Client start
Maps in snodeman
bool warning fix
SSEG Update
Show new number of SN's on add/remove in logs | Move output to debug level
sigTime
Temp Fix prior to major changes
SPORK Additions/Changes | SANDSTORM fee removal
prevent ganing sigTime too much on initial ssee
SandStorm/InstantX Functionality | Send with SandStorm on sendcoinsdialog.ui
Fix Filtering for active | add spaces to table-fy console view | add spaces to IP only in full mode
Merged Stormnode blinding
fix compiling for windows
fix console spam
Make stormnodes.dat verification more specific, prevent from overwriting corruped
files with unknown format, give more info in debug.log output
fixed shared key usage
fix even more console spam
rename stormnodes.dat->mncache.dat / fix strings
save nsqCount in mncache.dat
[Qt] add more NULL pointer checks in darksilkgui.cpp
close wallet regardless of "minimize to tray" settings
Added coincontrol sandstorm rounds column
Stormnode Proof of Service Added
Add structure for donating a percentage of ones earnings to another party via the protocol
Fixed snw message issue
fix pubkey/donation conditions
Fixed blinding submission process
increasing ss timeouts
fixed ssee sync
fixed Stormnode payments FindOldestNotInVec

src/qt/addresstablemodel.cpp
#include <QFont>
#include <QDebug>

#include "addresstablemodel.h"
#include "guiutil.h"
#include "walletmodel.h"
#include "wallet.h"
#include "base58.h"

const QString AddressTableModel::Send = "S";
const QString AddressTableModel::Receive = "R";

struct AddressTableEntry
{
    enum Type {
        Sending,
        Receiving
    };

    Type type;
    QString label;
    QString address;
    bool stealth;

    AddressTableEntry() {}
    AddressTableEntry(Type type, const QString &label, const QString &address, const bool &stealth = false):
        type(type), label(label), address(address), stealth(stealth) {}
};

struct AddressTableEntryLessThan
{
    bool operator()(const AddressTableEntry &a, const AddressTableEntry &b) const
    {
        return a.address < b.address;
    }
    bool operator()(const AddressTableEntry &a, const QString &b) const
    {
        return a.address < b;
    }
    bool operator()(const QString &a, const AddressTableEntry &b) const
    {
        return a < b.address;
    }
};

// Private implementation
class AddressTablePriv
{
public:
    CWallet *wallet;
    QList<AddressTableEntry> cachedAddressTable;
    AddressTableModel *parent;

    AddressTablePriv(CWallet *wallet, AddressTableModel *parent):
        wallet(wallet), parent(parent) {}

    void refreshAddressTable()
    {
        cachedAddressTable.clear();
        {
            LOCK(wallet->cs_wallet);
            BOOST_FOREACH(const PAIRTYPE(CTxDestination, std::string)& item, wallet->mapAddressBook)
            {
                const CDarkSilkAddress& address = item.first;
                const std::string& strName = item.second;
                bool fMine = IsMine(*wallet, address.Get());
                cachedAddressTable.append(AddressTableEntry(fMine ? AddressTableEntry::Receiving : AddressTableEntry::Sending,
                                  QString::fromStdString(strName),
                                  QString::fromStdString(address.ToString())));
            }

            std::set<CStealthAddress>::iterator it;
            for (it = wallet->stealthAddresses.begin(); it != wallet->stealthAddresses.end(); ++it)
            {
                bool fMine = !(it->scan_secret.size() < 1);
                cachedAddressTable.append(AddressTableEntry(fMine ? AddressTableEntry::Receiving : AddressTableEntry::Sending,
                                  QString::fromStdString(it->label),
                                  QString::fromStdString(it->Encoded()),
                                  true));
            };
        }
        // qLowerBound() and qUpperBound() require our cachedAddressTable list to be sorted in asc order
        qSort(cachedAddressTable.begin(), cachedAddressTable.end(), AddressTableEntryLessThan());
    }

    void updateEntry(const QString &address, const QString &label, bool isMine, int status)
    {
        // Find address / label in model
        QList<AddressTableEntry>::iterator lower = qLowerBound(
            cachedAddressTable.begin(), cachedAddressTable.end(), address, AddressTableEntryLessThan());
        QList<AddressTableEntry>::iterator upper = qUpperBound(
            cachedAddressTable.begin(), cachedAddressTable.end(), address, AddressTableEntryLessThan());
        int lowerIndex = (lower - cachedAddressTable.begin());
        int upperIndex = (upper - cachedAddressTable.begin());
        bool inModel = (lower != upper);
        AddressTableEntry::Type newEntryType = isMine ? AddressTableEntry::Receiving : AddressTableEntry::Sending;

        switch(status)
        {
        case CT_NEW:
            if(inModel)
            {
                qDebug() << "AddressTablePriv::updateEntry : Warning: Got CT_NEW, but entry is already in model";
                break;
            }
            parent->beginInsertRows(QModelIndex(), lowerIndex, lowerIndex);
            cachedAddressTable.insert(lowerIndex, AddressTableEntry(newEntryType, label, address));
            parent->endInsertRows();
            break;
        case CT_UPDATED:
            if(!inModel)
            {
                qDebug() << "AddressTablePriv::updateEntry : Warning: Got CT_UPDATED, but entry is not in model";
                break;
            }
            lower->type = newEntryType;
            lower->label = label;
            parent->emitDataChanged(lowerIndex);
            break;
        case CT_DELETED:
            if(!inModel)
            {
                qDebug() << "AddressTablePriv::updateEntry : Warning: Got CT_DELETED, but entry is not in model";
                break;
            }
            parent->beginRemoveRows(QModelIndex(), lowerIndex, upperIndex-1);
            cachedAddressTable.erase(lower, upper);
            parent->endRemoveRows();
            break;
        }
    }

    int size()
    {
        return cachedAddressTable.size();
    }

    AddressTableEntry *index(int idx)
    {
        if(idx >= 0 && idx < cachedAddressTable.size())
        {
            return &cachedAddressTable[idx];
        }
        else
        {
            return 0;
        }
    }
};

AddressTableModel::AddressTableModel(CWallet *wallet, WalletModel *parent) :
    QAbstractTableModel(parent),walletModel(parent),wallet(wallet),priv(0)
{
    columns << tr("Label") << tr("Address");
    priv = new AddressTablePriv(wallet, this);
    priv->refreshAddressTable();
}

AddressTableModel::~AddressTableModel()
{
    delete priv;
}

int AddressTableModel::rowCount(const QModelIndex &parent) const
{
    Q_UNUSED(parent);
    return priv->size();
}

int AddressTableModel::columnCount(const QModelIndex &parent) const
{
    Q_UNUSED(parent);
    return columns.length();
}

QVariant AddressTableModel::data(const QModelIndex &index, int role) const
{
    if(!index.isValid())
        return QVariant();

    AddressTableEntry *rec = static_cast<AddressTableEntry*>(index.internalPointer());

    if(role == Qt::DisplayRole || role == Qt::EditRole)
    {
        switch(index.column())
        {
        case Label:
            if(rec->label.isEmpty() && role == Qt::DisplayRole)
            {
                return tr("(no label)");
            }
            else
            {
                return rec->label;
            }
        case Address:
            return rec->address;
        }
    }
    else if (role == Qt::FontRole)
    {
        QFont font;
        if(index.column() == Address)
        {
            font = GUIUtil::darksilkAddressFont();
        }
        return font;
    }
    else if (role == TypeRole)
    {
        switch(rec->type)
        {
        case AddressTableEntry::Sending:
            return Send;
        case AddressTableEntry::Receiving:
            return Receive;
        default: break;
        }
    }
    return QVariant();
}

bool AddressTableModel::setData(const QModelIndex &index, const QVariant &value, int role)
{
    if(!index.isValid())
        return false;
    AddressTableEntry *rec = static_cast<AddressTableEntry*>(index.internalPointer());

    editStatus = OK;
    std::string strTemp, strValue;
    if(role == Qt::EditRole)
    {
        switch(index.column())
        {
        case Label:
            // Do nothing, if old label == new label
            if(rec->label == value.toString())
            {
                editStatus = NO_CHANGES;
                return false;
            }

            strTemp = rec->address.toStdString();
            if (IsStealthAddress(strTemp))
            {
                strValue = value.toString().toStdString();
                wallet->UpdateStealthAddress(strTemp, strValue, false);
            } else
            {
                wallet->SetAddressBookName(CDarkSilkAddress(strTemp).Get(), value.toString().toStdString());
            }
            break;
        case Address:
            std::string sTemp = value.toString().toStdString();
            if (IsStealthAddress(sTemp))
            {
                printf("TODO\n");
                editStatus = INVALID_ADDRESS;
                return false;
            }
            // Do nothing, if old address == new address
            if(CDarkSilkAddress(rec->address.toStdString()) == CDarkSilkAddress(value.toString().toStdString()))
            {
                editStatus = NO_CHANGES;
                return false;
            }
            // Refuse to set invalid address, set error status and return false
            else if(!walletModel->validateAddress(value.toString()))
            {
                editStatus = INVALID_ADDRESS;
                return false;
            }
            // Check for duplicate addresses to prevent accidental deletion of addresses, if you try
            // to paste an existing address over another address (with a different label)
            else if(wallet->mapAddressBook.count(CDarkSilkAddress(value.toString().toStdString()).Get()))
            {
                editStatus = DUPLICATE_ADDRESS;
                return false;
            }
            // Double-check that we're not overwriting a receiving address
            else if(rec->type == AddressTableEntry::Sending)
            {
                {
                    LOCK(wallet->cs_wallet);
                    // Remove old entry
                    wallet->DelAddressBookName(CDarkSilkAddress(rec->address.toStdString()).Get());
                    // Add new entry with new address
                    wallet->SetAddressBookName(CDarkSilkAddress(value.toString().toStdString()).Get(), rec->label.toStdString());
                }
            }
            break;
        }
        return true;
    }
    return false;
}

QVariant AddressTableModel::headerData(int section, Qt::Orientation orientation, int role) const
{
    if(orientation == Qt::Horizontal)
    {
        if(role == Qt::DisplayRole)
        {
            return columns[section];
        }
    }
    return QVariant();
}

Qt::ItemFlags AddressTableModel::flags(const QModelIndex &index) const
{
    if(!index.isValid())
        return 0;
    AddressTableEntry *rec = static_cast<AddressTableEntry*>(index.internalPointer());

    Qt::ItemFlags retval = Qt::ItemIsSelectable | Qt::ItemIsEnabled;
    // Can edit address and label for sending addresses,
    // and only label for receiving addresses.
    if(rec->type == AddressTableEntry::Sending ||
      (rec->type == AddressTableEntry::Receiving && index.column()==Label))
    {
        retval |= Qt::ItemIsEditable;
    }
    return retval;
}

QModelIndex AddressTableModel::index(int row, int column, const QModelIndex &parent) const
{
    Q_UNUSED(parent);
    AddressTableEntry *data = priv->index(row);
    if(data)
    {
        return createIndex(row, column, priv->index(row));
    }
    else
    {
        return QModelIndex();
    }
}

void AddressTableModel::updateEntry(const QString &address, const QString &label, bool isMine, int status)
{
    // Update address book model from DarkSilk core
    priv->updateEntry(address, label, isMine, status);
}

QString AddressTableModel::addRow(const QString &type, const QString &label, const QString &address, int addressType)
{
    std::string strLabel = label.toStdString();
    std::string strAddress = address.toStdString();

    editStatus = OK;

    if(type == Send)
    {
        if (strAddress.length() > 75)
        {
            CStealthAddress sxAddr;
            if (!sxAddr.SetEncoded(strAddress))
            {
                editStatus = INVALID_ADDRESS;
                return QString();
            }
            
            // -- Check for duplicate addresses
            {
                LOCK(wallet->cs_wallet);
                
                if (wallet->stealthAddresses.count(sxAddr))
                {
                    editStatus = DUPLICATE_ADDRESS;
                    return QString();
                };
                
                sxAddr.label = strLabel;
                wallet->AddStealthAddress(sxAddr);
            }
            
        } else
        {
            if (!walletModel->validateAddress(address))
            {
                editStatus = INVALID_ADDRESS;
                return QString();
            };
            // Check for duplicate addresses
            {
                LOCK(wallet->cs_wallet);
                if (wallet->mapAddressBook.count(CDarkSilkAddress(strAddress).Get()))
                {
                    editStatus = DUPLICATE_ADDRESS;
                    return QString();
                };
                
                wallet->SetAddressBookName(CDarkSilkAddress(strAddress).Get(), strLabel);
            }
        }
    }
    else if(type == Receive)
    {
        // Generate a new address to associate with given label
        WalletModel::UnlockContext ctx(walletModel->requestUnlock());
        
        if(!ctx.isValid())
        {
            // Unlock wallet failed or was cancelled
            editStatus = WALLET_UNLOCK_FAILURE;
            return QString();
        }
        
        if (addressType == AT_Stealth)
        {
            CStealthAddress newStealthAddr;
            std::string sError;
            if (!wallet->NewStealthAddress(sError, strLabel, newStealthAddr)
                || !wallet->AddStealthAddress(newStealthAddr))
            {
                editStatus = KEY_GENERATION_FAILURE;
                return QString();
            }
            strAddress = newStealthAddr.Encoded();
        } else
        {
            CPubKey newKey;
            if(!wallet->GetKeyFromPool(newKey))
            {
                editStatus = KEY_GENERATION_FAILURE;
                return QString();
            }
            strAddress = CDarkSilkAddress(newKey.GetID()).ToString();
            
            {
                LOCK(wallet->cs_wallet);
                wallet->SetAddressBookName(CDarkSilkAddress(strAddress).Get(), strLabel);
            }
        }
    }
    else
    {
        return QString();
    }

    
    return QString::fromStdString(strAddress);
}

bool AddressTableModel::removeRows(int row, int count, const QModelIndex &parent)
{
    Q_UNUSED(parent);
    AddressTableEntry *rec = priv->index(row);
    if(count != 1 || !rec || rec->type == AddressTableEntry::Receiving)
    {
        // Can only remove one row at a time, and cannot remove rows not in model.
        // Also refuse to remove receiving addresses.
        return false;
    }
    {
        LOCK(wallet->cs_wallet);
        wallet->DelAddressBookName(CDarkSilkAddress(rec->address.toStdString()).Get());
    }
    return true;
}

/* Look up label for address in address book, if not found return empty string.
 */
QString AddressTableModel::labelForAddress(const QString &address) const
{
    {
        LOCK(wallet->cs_wallet);
        std::string sAddr = address.toStdString();
        
        if (sAddr.length() > 75)
        {
            CStealthAddress sxAddr;
            if (!sxAddr.SetEncoded(sAddr))
                return QString();
            
            std::set<CStealthAddress>::iterator it;
            it = wallet->stealthAddresses.find(sxAddr);
            if (it == wallet->stealthAddresses.end())
                return QString();
            
            return QString::fromStdString(it->label);
        } else
        {
            CDarkSilkAddress address_parsed(sAddr);
            std::map<CTxDestination, std::string>::iterator mi = wallet->mapAddressBook.find(address_parsed.Get());
            if (mi != wallet->mapAddressBook.end())
            {
                return QString::fromStdString(mi->second);
            }
        }
    }
    return QString();
}

int AddressTableModel::lookupAddress(const QString &address) const
{
    QModelIndexList lst = match(index(0, Address, QModelIndex()),
                                Qt::EditRole, address, 1, Qt::MatchExactly);
    if(lst.isEmpty())
    {
        return -1;
    }
    else
    {
        return lst.at(0).row();
    }
}

void AddressTableModel::emitDataChanged(int idx)
{
    emit dataChanged(index(idx, 0, QModelIndex()), index(idx, columns.length()-1, QModelIndex()));
}
